sources/api/netdog/src/net_config/v3.rs

```diff
@@ -0,0 +1,106 @@
+//! The `v3` module contains the third version of the network configuration and implements the
+//! appropriate traits.
+
+use super::devices::NetworkDeviceV1;
+use super::{error, Interfaces, Result, Validate};
+use crate::interface_name::InterfaceName;
+use crate::wicked::{WickedInterface, WickedLinkConfig};
+use indexmap::IndexMap;
+use serde::Deserialize;
+use snafu::ensure;
+
+#[derive(Debug, Deserialize)]
+pub(crate) struct NetConfigV3 {
+    #[serde(flatten)]
+    pub(crate) net_devices: IndexMap<InterfaceName, NetworkDeviceV1>,
+}
+
+impl Interfaces for NetConfigV3 {
+    fn primary_interface(&self) -> Option<String> {
+        self.net_devices
+            .iter()
+            .find(|(_, v)| v.primary() == Some(true))
+            .or_else(|| self.net_devices.first())
+            .map(|(n, _)| n.to_string())
+    }
+
+    fn has_interfaces(&self) -> bool {
+        !self.net_devices.is_empty()
+    }
+
+    fn as_wicked_interfaces(&self) -> Vec<WickedInterface> {
+        let mut wicked_interfaces = Vec::new();
+        for (name, config) in &self.net_devices {
+            let interface = WickedInterface::from((name, config));
+
+            // If config is a Bond, we will generate the interface configuration. For each interface:
+            // WickedInterface::new(name), add that to wicked_interfaces
+            if let NetworkDeviceV1::BondDevice(b) = config {
+                for device in &b.interfaces {
+                    let mut wicked_sub_interface = WickedInterface::new(device.clone());
+                    wicked_sub_interface.link = Some(WickedLinkConfig {
+                        master: name.clone(),
+                    });
+
+                    wicked_interfaces.push(wicked_sub_interface)
+                }
+            }
+
+            wicked_interfaces.push(interface)
+        }
+
+        wicked_interfaces
+    }
+}
+
+#[allow(clippy::to_string_in_format_args)]
+impl Validate for NetConfigV3 {
+    fn validate(&self) -> Result<()> {
+        for (name, device) in &self.net_devices {
+            // Bonds create the interfaces automatically, specifying those interfaces would cause a
+            // collision so this emits an error
+            if let NetworkDeviceV1::BondDevice(config) = device {
+                for interface in &config.interfaces {
+                    if self.net_devices.contains_key(interface) {
+                        return error::InvalidNetConfigSnafu {
+                            reason: format!(
+                                "{} in bond {} cannot be manually configured",
+                                interface.to_string(),
+                                name.to_string()
+                            ),
+                        }
+                        .fail();
```

Comment:
I don't believe so. But this logic wouldn't catch that, it only catches if a user specified the raw physical interface directly. If they specify 
```toml
[bond0]
kind = bond,
interfaces = ["eno1", "eno2"]
....

[tag1]
kind = vlan
device = "eno1"
....
```
We wouldn't catch them both pointing at it. We'd only catch if they specified `eno1` directly like so:
```toml
[eno1]
dhcp4 = true
```
Its a little more tricky to do this deep validation but I can add it (after I confirm that truly won't work).

Reply:
After I wrote this, it occurred to me that I could create a `HashSet` of all the names and simply check for them there. It forces another loop over all the devices but from here we could do other checks if we need. I'm open to improvements since we are now looping twice but I couldn't seem to find a good way to ensure bonds don't refer to devices in vlans if we haven't found the conflicting vlan yet in the loop. In this case, we don't expect many interfaces to be configured so looping twice shouldn't be too high of a penalty to catch this edge case.